README.md
# pybic

This is a driver for Meanwell bi-directional charger.

## Author
- Frank Chang

## License
See LICENSE


CHANGELOG.md
# CHANGELOG



## v0.3.0 (2023-09-19)

### Feature

* feat: support reading iout_set and vout_set ([`1d03f20`](https://github.com/frankurcrazy/pybic/commit/1d03f206214ac2c524353735c65600ab917ea5cd))


## v0.2.0 (2023-09-01)

### Feature

* feat(bic): add support for system_status

fix(bic): fix incorrect fault_status mapping ([`4f4a00a`](https://github.com/frankurcrazy/pybic/commit/4f4a00a4163f17af8e7fe210e14960ab1d255d90))


## v0.1.3 (2023-09-01)

### Fix

* fix(listener): fix hanging issue described in 1cf8f3c

the hanging described in previous commit is due to deadlock in
can.Notifier, essentially you should not block
`Notifier.on_message_received` as it is called after acquiring a mutex,
so that only one handler may be executed at a time.

the solution to this problem is to execute handler asynchrounusly
with a thread pool. ([`220f733`](https://github.com/frankurcrazy/pybic/commit/220f733b2473980567bd6b2dcf53e0acc9a8a3da))


## v0.1.2 (2023-09-01)

### Fix

* fix(bic): prefetch scaling factor to avoid hanging

fix(bic): strip mfr model

when quering temperature without first quering scaling factor,
the library would hang, without receving scaling factor response.

and no subsequent response is ever received until a system reboot.
by prefetching scaling factor before doing anything, we can workaround
the issue. ([`1cf8f3c`](https://github.com/frankurcrazy/pybic/commit/1cf8f3c1413117d50a20ac5a02d7a77a01602c8e))


## v0.1.1 (2023-08-31)

### Chore

* chore: bump poetry.lock ([`7ab14a5`](https://github.com/frankurcrazy/pybic/commit/7ab14a5b9099328404206c440a41481e089989ee))

### Fix

* fix: expose Bic to the pybic package ([`cb68adf`](https://github.com/frankurcrazy/pybic/commit/cb68adf69d4204c16486b5aa789813e544174bf8))


## v0.1.0 (2023-08-31)

### Chore

* chore: initial commit ([`99c4adb`](https://github.com/frankurcrazy/pybic/commit/99c4adbd6e697d2e8221af37ed1046d27da8afa2))

### Feature

* feat(bic): add support for more bic commands

fix(listener): filter message not meant for the controller

Added support for:
- temperature (r)
- fault_status (r)
- mfr_id (r)
- mfr_model (r)
- mfr_serial (r)
- mfr_date (r)
- mfr_revision (r)
- mfr_location (r) ([`9ce39df`](https://github.com/frankurcrazy/pybic/commit/9ce39dff0fa474f922c4b11eef1d380389d9a4c2))

* feat(bic): impl. bic property getters and setters

feat(listener): impl. CAN listener for bic
feat(utils): impl. can message formatter

Supported Properties:
- operation (r/w)
- v_in (r)
- v_out (r/w)
- i_out (r/w)
- reverse_v_out (r/w)
- reverse_i_out (r/w)
- scaling_factor (r)
- bidirectional_config (r/w)
- system_config (r/w)
- direction_ctrl (r/w) ([`f91c54f`](https://github.com/frankurcrazy/pybic/commit/f91c54f2b2f4b95e1b540d0c8856e390a68c90e4))

### Refactor

* refactor(promise): use promise for asynchronus results ([`c9c8d26`](https://github.com/frankurcrazy/pybic/commit/c9c8d2686dd4f09f1deb78751dac6545ee43b493))

### Style

* style: reformat code with black, isort ([`f1c3c82`](https://github.com/frankurcrazy/pybic/commit/f1c3c8249bbcb3c727b41cbaac09ad6207294682))


tests/__init__.py
import pathlib
import sys

# add source path
sys.path.insert(0, pathlib.Path().joinpath("src").absolute().__str__())

import pybic


tests/test_import_pybic.py
import unittest


class TestImportPyBic(unittest.TestCase):
    def test_import_pybic(self):
        try:
            import pybic
        except Exception as e:
            self.fail(f"failed importing pybic: {e}")


src/pybic/cmd.py
SYSTEM_CONFIG_OPERATION_INIT_OFF = 0x00
SYSTEM_CONFIG_OPERATION_INIT_ON = 0x01
SYSTEM_CONFIG_OPERATION_INIT_LAST = 0x02
SYSTEM_CONFIG_CAN_CTRL_SVR = 0x00
SYSTEM_CONFIG_CAN_CTRL_CAN = 0x01

BIDIRECTIONAL_CONFIG_MODE_CV = 0x00
BIDIRECTIONAL_CONFIG_MODE_BATTERY = 0x01

DIRECTION_CTRL_CHARGE = 0x00
DIRECTION_CTRL_DISCHARGE = 0x01

OPERATION = (0x0000, "rw", 1)
VOUT_SET = (0x0020, "rw", 2)
IOUT_SET = (0x0030, "rw", 2)
FAULT_STATUS = (0x0040, "r", 2)
READ_VIN = (0x0050, "r", 2)
READ_VOUT = (0x0060, "r", 2)
READ_IOUT = (0x0061, "r", 2)
READ_TEMPERATURE = (0x0062, "r", 2)
MFR_ID_B0B5 = (0x0080, "r", 6)
MFR_ID_B6B11 = (0x0081, "r", 6)
MFR_MODEL_B0B5 = (0x0082, "r", 6)
MFR_MODEL_B6B11 = (0x0083, "r", 6)
MFR_REVISION_B0B5 = (0x0084, "r", 6)
MFR_LOCATION_B0B2 = (0x0085, "rw", 3)
MFR_DATE_B0B5 = (0x0086, "rw", 6)
MFR_SERIAL_B0B5 = (0x0087, "rw", 6)
MFR_SERIAL_B6B11 = (0x0088, "rw", 6)
SCALING_FACTOR = (0x00C0, "r", 2)
SYSTEM_STATUS = (0x00C1, "r", 2)
SYSTEM_CONFIG = (0x00C2, "rw", 2)
DIRECTION_CTRL = (0x0100, "rw", 1)
REVERSE_VOUT_SET = (0x0120, "rw", 2)
REVERSE_IOUT_SET = (0x0130, "rw", 2)
BIDIRECTIONAL_CONFIG = (0x0140, "rw", 2)

CMD_LOOKUP_TABLE = {
    0x0000: OPERATION,
    0x0020: VOUT_SET,
    0x0030: IOUT_SET,
    0x0040: FAULT_STATUS,
    0x0050: READ_VIN,
    0x0060: READ_VOUT,
    0x0061: READ_IOUT,
    0x0062: READ_TEMPERATURE,
    0x0080: MFR_ID_B0B5,
    0x0081: MFR_ID_B6B11,
    0x0082: MFR_MODEL_B0B5,
    0x0083: MFR_MODEL_B6B11,
    0x0084: MFR_REVISION_B0B5,
    0x0085: MFR_LOCATION_B0B2,
    0x0086: MFR_DATE_B0B5,
    0x0087: MFR_SERIAL_B0B5,
    0x0088: MFR_SERIAL_B6B11,
    0x00C0: SCALING_FACTOR,
    0x00C1: SYSTEM_STATUS,
    0x00C2: SYSTEM_CONFIG,
    0x0100: DIRECTION_CTRL,
    0x0120: REVERSE_VOUT_SET,
    0x0130: REVERSE_IOUT_SET,
    0x0140: BIDIRECTIONAL_CONFIG,
}


src/pybic/listener.py
import atexit
import concurrent.futures
import logging
import struct

import can

import pybic.cmd
from pybic.utils import format_can_message

DIR_TO_CONTROLLER = 0xC0200
DIR_FROM_CONTROLLER = 0xC0300
CMD_LENGTH = 2


def convert_scaling_factor(sf):
    if sf == 0:
        return None

    return 10 ** (sf - 7)


class BicListener(can.Listener):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._logger = logging.getLogger(__name__)
        self._bic = kwargs.get("bic", None)

        self._handler_thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=2)

        self._command_handlers = {
            pybic.cmd.OPERATION[0]: self._handle_cmd_operation,
            pybic.cmd.SCALING_FACTOR[0]: self._handle_cmd_scaling_factor,
            pybic.cmd.READ_VIN[0]: self._handle_cmd_read_vin,
            pybic.cmd.READ_IOUT[0]: self._handle_cmd_read_iout,
            pybic.cmd.REVERSE_VOUT_SET[0]: self._handle_cmd_reverse_vout,
            pybic.cmd.READ_VOUT[0]: self._handle_cmd_read_vout,
            pybic.cmd.VOUT_SET[0]: self._handle_cmd_vout_set,
            pybic.cmd.REVERSE_IOUT_SET[0]: self._handle_cmd_reverse_iout,
            pybic.cmd.IOUT_SET[0]: self._handle_cmd_iout_set,
            pybic.cmd.SYSTEM_CONFIG[0]: self._handle_cmd_system_config,
            pybic.cmd.BIDIRECTIONAL_CONFIG[0]: self._handle_cmd_bidirectional_config,
            pybic.cmd.DIRECTION_CTRL[0]: self._handle_cmd_direction_ctrl,
            pybic.cmd.READ_TEMPERATURE[0]: self._handle_cmd_read_temperature,
            pybic.cmd.FAULT_STATUS[0]: self._handle_cmd_fault_status,
            pybic.cmd.SYSTEM_STATUS[0]: self._handle_cmd_system_status,
            pybic.cmd.MFR_ID_B0B5[0]: self._handle_cmd_mfr_id_b0b5,
            pybic.cmd.MFR_ID_B6B11[0]: self._handle_cmd_mfr_id_b6b11,
            pybic.cmd.MFR_MODEL_B0B5[0]: self._handle_cmd_mfr_model_b0b5,
            pybic.cmd.MFR_MODEL_B6B11[0]: self._handle_cmd_mfr_model_b6b11,
            pybic.cmd.MFR_SERIAL_B0B5[0]: self._handle_cmd_mfr_serial_b0b5,
            pybic.cmd.MFR_SERIAL_B6B11[0]: self._handle_cmd_mfr_serial_b6b11,
            pybic.cmd.MFR_DATE_B0B5[0]: self._handle_cmd_mfr_date_b0b5,
            pybic.cmd.MFR_REVISION_B0B5[0]: self._handle_cmd_mfr_revision_b0b5,
            pybic.cmd.MFR_LOCATION_B0B2[0]: self._handle_cmd_mfr_location_b0b2,
        }

        if not self._bic:
            self._logger.warn(f"bic is none")

        atexit.register(self.stop)

    def on_error(self, exc: Exception) -> None:
        self._logger.error(f"encounter error: {exc}")

    def on_message_received(self, msg: can.message.Message) -> None:
        try:
            if (msg.arbitration_id & 0xFFF00) == DIR_TO_CONTROLLER:
                self._logger.debug(
                    f"received message to controller: {format_can_message(msg)}"
                )
                self._handle_to_controller(msg)

            if (msg.arbitration_id & 0xFFF00) == DIR_FROM_CONTROLLER:
                self._logger.debug(
                    f"received message from controller: {format_can_message(msg)}"
                )
                self._handle_from_controller(msg)

        except Exception as e:
            self._logger.error(
                f'failed handling message "{format_can_message(msg)}": {e}'
            )

    def stop(self) -> None:
        self._handler_job_queue.shutdown(cancel_futures=True)

    def _fulfill_bic_promises(self, key, value):
        if not self._bic:
            self._logger.warn(f"bic is none, skip fulfilling promises for key {key}")
            return

        if key not in self._bic._promises:
            self._logger.warn(
                f"key does not exist in promise map, skip fulfilling promises for key {key}"
            )
            return

        for promise in self._bic._promises[key]:
            promise.set_result(value)

        self._bic._promises[key].clear()

    def _handle_to_controller(self, msg: can.message.Message) -> None:
        if msg.arbitration_id & 0x000FF != self._bic._arbitration_id & 0x000FF:
            # skip message not meant for us
            return

        (cmd,) = struct.unpack("<H", msg.data[:CMD_LENGTH])

        if cmd not in self._command_handlers:
            self._logger.warn(f"handler not implemented for command: {cmd}")
            return

        self._handler_thread_pool.submit(self._command_handlers[cmd], msg=msg)

    def _handle_from_controller(self, msg: can.message.Message) -> None:
        pass

    def _handle_cmd_operation(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("operation", True if msg.data[2] == 1 else False)

    def _handle_cmd_read_vin(self, msg: can.message.Message) -> None:
        (v_in_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises("v_in", v_in_raw * self._bic.scaling_factor["v_in"])

    def _handle_cmd_read_vout(self, msg: can.message.Message) -> None:
        (v_out_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "v_out", v_out_raw * self._bic.scaling_factor["v_out"]
        )

    def _handle_cmd_vout_set(self, msg: can.message.Message) -> None:
        (v_out_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "v_out_set", v_out_raw * self._bic.scaling_factor["v_out"]
        )

    def _handle_cmd_scaling_factor(self, msg: can.message.Message) -> None:
        # trim the command from the data
        data = msg.data[2:]

        v_out_factor = convert_scaling_factor(data[0] & 0x0F)
        i_out_factor = convert_scaling_factor((data[0] & 0xF0) >> 4)
        v_in_factor = convert_scaling_factor(data[1] & 0x0F)
        fan_speed_factor = convert_scaling_factor((data[1] & 0xF0) >> 4)
        temperature_factor = convert_scaling_factor(data[2] & 0x0F)
        i_in_factor = convert_scaling_factor(data[3] & 0x0F)

        self._fulfill_bic_promises(
            "scaling_factor",
            {
                "v_out": v_out_factor,
                "i_out": i_out_factor,
                "v_in": v_in_factor,
                "i_in": i_in_factor,
                "fan_speed": fan_speed_factor,
                "temperature": temperature_factor,
            },
        )

    def _handle_cmd_system_config(self, msg: can.message.Message) -> None:
        # trim the command from the data
        data = msg.data[2:]

        operation_init = (data[0] & 0x06) >> 1
        can_ctrl = data[0] & 0x01

        self._fulfill_bic_promises(
            "system_config",
            {
                "operation_init": operation_init,
                "can_ctrl": can_ctrl,
            },
        )

    def _handle_cmd_bidirectional_config(self, msg: can.message.Message) -> None:
        # trim the command from the data
        data = msg.data[2:]

        self._fulfill_bic_promises(
            "bidirectional_config",
            {
                "mode": data[0] & 0x01,
            },
        )

    def _handle_cmd_direction_ctrl(self, msg: can.message.Message) -> None:
        # trim the command from the data
        data = msg.data[2]

        self._fulfill_bic_promises("direction_ctrl", data)

    def _handle_cmd_reverse_vout(self, msg: can.message.Message) -> None:
        (rev_v_out_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "reverse_v_out", rev_v_out_raw * self._bic.scaling_factor["v_out"]
        )

    def _handle_cmd_reverse_iout(self, msg: can.message.Message) -> None:
        (rev_i_out_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "reverse_i_out", rev_i_out_raw * self._bic.scaling_factor["i_out"]
        )

    def _handle_cmd_iout_set(self, msg: can.message.Message) -> None:
        (i_out_set_raw,) = struct.unpack("<H", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "i_out_set", i_out_set_raw * self._bic.scaling_factor["i_out"]
        )

    def _handle_cmd_read_iout(self, msg: can.message.Message) -> None:
        (i_out_raw,) = struct.unpack("<h", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "i_out", i_out_raw * self._bic.scaling_factor["i_out"]
        )

    def _handle_cmd_read_temperature(self, msg: can.message.Message) -> None:
        (temp_raw,) = struct.unpack("<h", msg.data[2 : msg.dlc])

        self._fulfill_bic_promises(
            "temperature", temp_raw * self._bic.scaling_factor["temperature"]
        )

    def _handle_cmd_fault_status(self, msg: can.message.Message) -> None:
        data = msg.data[2 : msg.dlc]

        fault_status = {
            "fan_fail": (data[0] & 0x01) > 0,
            "otp": (data[0] & 0x02) > 0,
            "ovp": (data[0] & 0x04) > 0,
            "olp": (data[0] & 0x08) > 0,
            "short": (data[0] & 0x10) > 0,
            "ac_fail": (data[0] & 0x20) > 0,
            "op_off": (data[0] & 0x40) > 0,
            "hi_temp": (data[0] & 0x80) > 0,
            "hv_ovp": (data[1] & 0x01) > 0,
        }

        self._fulfill_bic_promises("fault_status", fault_status)

    def _handle_cmd_system_status(self, msg: can.message.Message) -> None:
        data = msg.data[2 : msg.dlc]

        system_status = {
            "m/s": (data[0] & 0x01) > 0,
            "dc_ok": (data[0] & 0x02) > 0,
            "pfc_ok": (data[0] & 0x04) > 0,
            "adl_on": (data[0] & 0x10) > 0,
            "initial_state": (data[0] & 0x20) > 0,
            "eeper": (data[0] & 0x40) > 0,
        }

        self._fulfill_bic_promises("system_status", system_status)

    def _handle_cmd_mfr_id_b0b5(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_id_b0b5", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_id_b6b11(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_id_b6b11", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_model_b0b5(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_model_b0b5", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_model_b6b11(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_model_b6b11", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_serial_b0b5(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_serial_b0b5", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_serial_b6b11(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_serial_b6b11", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_date_b0b5(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_date_b0b5", msg.data[2 : msg.dlc].decode())

    def _handle_cmd_mfr_revision_b0b5(self, msg: can.message.Message) -> None:
        revision_list = []
        for rev in msg.data[2 : msg.dlc]:
            if rev != 0xFF:
                revision_list.append(f"r{rev/10:2.1f}")

        self._fulfill_bic_promises("mfr_revision_b0b5", tuple(revision_list))

    def _handle_cmd_mfr_location_b0b2(self, msg: can.message.Message) -> None:
        self._fulfill_bic_promises("mfr_location_b0b2", msg.data[2 : msg.dlc].decode())


src/pybic/bic.py
import atexit
import collections
import functools
import logging
import math
import queue
import struct
import threading
import time

import can

import pybic.cmd
import pybic.listener
from pybic.utils import Promise, format_can_message


def wait_for(cmd):
    def decorator(method):
        def _impl(self, *args, **kwargs):
            if cmd not in self._promises:
                self._promises[cmd] = collections.deque()
            promise = Promise()
            self._promises[cmd].append(promise)
            method(self, *args, *kwargs)

            return promise.result

        return _impl

    return decorator


class MessageSender:
    """Command sender queues the command sending request and create a task
    to process the command to be sent
    """

    def __init__(self, bus: can.Bus):
        self._stop = threading.Event()
        self._queue = queue.Queue(maxsize=16)
        self._logger = logging.getLogger(__name__)
        self._bus = bus

        self._thread = threading.Thread(
            target=self._worker,
        )
        self._thread.start()

        atexit.register(self.stop)

    def _worker(self) -> None:
        self._logger.info("starting worker")
        while not self._stop.is_set():
            try:
                msg = self._queue.get_nowait()
                self._bus.send(msg)
                self._queue.task_done()
                self._logger.debug(f"message {format_can_message(msg)} sent.")
            except queue.Empty:
                time.sleep(0.1)
                continue

        self._logger.info("worker stopped")

    def send(self, msg: can.message.Message) -> None:
        self._queue.put(msg)

    def stop(self):
        self._logger.info(f"stopping message sender")
        self._stop.set()
        self._thread.join()


class Bic:
    def __init__(self, bus: can.Bus, arbitration_id: int = 0xC0300):
        self._bus = bus
        self._arbitration_id = arbitration_id
        self._msg_sender = MessageSender(bus=bus)
        self._listener = pybic.listener.BicListener(bic=self)
        can.Notifier(bus, [self._listener])

        self._promises = {}

        # prefetch scaling factor
        self.scaling_factor

    def _msg_factory(self, command, data=None) -> can.message.Message:
        data_bytes = struct.pack("<H", command[0])

        if data:
            data_bytes += data

        return can.Message(
            arbitration_id=self._arbitration_id, is_extended_id=True, data=data_bytes
        )

    @property
    @wait_for("operation")
    def operation(self):
        """query the operation of bic"""

        msg = self._msg_factory(command=pybic.cmd.OPERATION)
        self._msg_sender.send(msg)

    @operation.setter
    def operation(self, value: bool) -> None:
        """set the operation of bic

        arguments:
            value: 1 for on, 0 for off
        """

        data = b"\x01" if value else b"\x00"
        msg = self._msg_factory(command=pybic.cmd.OPERATION, data=data)
        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("scaling_factor")
    def scaling_factor(self):
        """query the scaling factor of the bic"""
        msg = self._msg_factory(command=pybic.cmd.SCALING_FACTOR)
        self._msg_sender.send(msg)

    @property
    @wait_for("v_in")
    def v_in(self):
        """query the v_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.READ_VIN)

        self._msg_sender.send(msg)

    @property
    @wait_for("reverse_v_out")
    def reverse_v_out(self):
        """query the reverse_v_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.REVERSE_VOUT_SET)

        self._msg_sender.send(msg)

    @reverse_v_out.setter
    def reverse_v_out(self, value):
        """set the reverse_v_out of the bic"""
        if value < 38 or value > 65:
            raise ValueError("reverse_v_out should be [38, 65]")

        rev_vout_scaled = math.floor(value / self.scaling_factor["v_out"])
        rev_vout_bytes = struct.pack("<H", rev_vout_scaled)

        msg = self._msg_factory(command=pybic.cmd.REVERSE_VOUT_SET, data=rev_vout_bytes)

        self._msg_sender.send(msg)

    @property
    @wait_for("reverse_i_out")
    def reverse_i_out(self):
        """query the reverse_i_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.REVERSE_IOUT_SET)

        self._msg_sender.send(msg)

    @reverse_i_out.setter
    def reverse_i_out(self, value):
        """set the reverse_i_out of the bic"""
        if value < 0.45 or value > 38.3:
            raise ValueError("reverse_i_out should be [0.45, 38.3]")

        rev_iout_scaled = math.floor(value / self.scaling_factor["i_out"])
        rev_iout_bytes = struct.pack("<H", rev_iout_scaled)

        msg = self._msg_factory(command=pybic.cmd.REVERSE_IOUT_SET, data=rev_iout_bytes)

        self._msg_sender.send(msg)

    @property
    @wait_for("v_out")
    def v_out(self):
        """query the v_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.READ_VOUT)

        self._msg_sender.send(msg)

    @property
    @wait_for("v_out_set")
    def v_out_set(self):
        """query the v_out_set of the bic"""
        msg = self._msg_factory(command=pybic.cmd.VOUT_SET)

        self._msg_sender.send(msg)

    @v_out.setter
    def v_out(self, value):
        """set the v_out of the bic"""
        if value < 38 or value > 65:
            raise ValueError("v_out should be [38, 65]")

        vout_scaled = math.floor(value / self.scaling_factor["v_out"])
        vout_bytes = struct.pack("<H", vout_scaled)

        msg = self._msg_factory(command=pybic.cmd.VOUT_SET, data=vout_bytes)

        self._msg_sender.send(msg)

    @property
    @wait_for("i_out")
    def i_out(self):
        """query the i_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.READ_IOUT)

        self._msg_sender.send(msg)

    @property
    @wait_for("i_out_set")
    def i_out_set(self):
        """query the i_out of the bic"""
        msg = self._msg_factory(command=pybic.cmd.IOUT_SET)

        self._msg_sender.send(msg)

    @i_out.setter
    def i_out(self, value):
        """set the i_out of the bic"""
        if value < 0.45 or value > 49.5:
            raise ValueError("i_out should be [0.45, 49.5]")

        vout_scaled = math.floor(value / self.scaling_factor["i_out"])
        vout_bytes = struct.pack("<H", vout_scaled)

        msg = self._msg_factory(command=pybic.cmd.IOUT_SET, data=vout_bytes)

        self._msg_sender.send(msg)

    @property
    @wait_for("system_config")
    def system_config(self):
        msg = self._msg_factory(command=pybic.cmd.SYSTEM_CONFIG)

        self._msg_sender.send(msg)

    @system_config.setter
    def system_config(self, value: dict) -> None:
        if "can_ctrl" not in value:
            raise ValueError("can_ctrl key not found")
        if "operation_init" not in value:
            raise ValueError("operation_init key not found")

        data = bytes(
            [(value["can_ctrl"] & 0x01) + ((value["operation_init"] & 0x06) << 1), 0x00]
        )

        msg = self._msg_factory(command=pybic.cmd.SYSTEM_CONFIG, data=data)

        self._msg_sender.send(msg)

    @property
    @wait_for("bidirectional_config")
    def bidirectional_config(self):
        msg = self._msg_factory(command=pybic.cmd.BIDIRECTIONAL_CONFIG)

        self._msg_sender.send(msg)

    @bidirectional_config.setter
    def bidirectional_config(self, value: dict) -> None:
        if "mode" not in value:
            raise ValueError("mode key not found")

        msg = self._msg_factory(
            command=pybic.cmd.BIDIRECTIONAL_CONFIG, data=bytes([value["mode"], 0x00])
        )

        self._msg_sender.send(msg)

    @property
    @wait_for("direction_ctrl")
    def direction_ctrl(self):
        msg = self._msg_factory(command=pybic.cmd.DIRECTION_CTRL)

        self._msg_sender.send(msg)

    @direction_ctrl.setter
    def direction_ctrl(self, value: int) -> None:
        msg = self._msg_factory(
            command=pybic.cmd.DIRECTION_CTRL,
            data=bytes(
                [
                    value,
                ]
            ),
        )

        self._msg_sender.send(msg)

    @property
    @wait_for("temperature")
    def temperature(self):
        """query the temperature of the bic"""
        msg = self._msg_factory(command=pybic.cmd.READ_TEMPERATURE)

        self._msg_sender.send(msg)

    @property
    @wait_for("fault_status")
    def fault_status(self):
        """query the fault status of the bic"""
        msg = self._msg_factory(command=pybic.cmd.FAULT_STATUS)

        self._msg_sender.send(msg)

    @property
    @wait_for("system_status")
    def system_status(self):
        """query the system status of the bic"""
        msg = self._msg_factory(command=pybic.cmd.SYSTEM_STATUS)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_id_b0b5")
    def mfr_id_b0b5(self):
        """query the mfr_id_b0b5 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_ID_B0B5)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_id_b6b11")
    def mfr_id_b6b11(self):
        """query the mfr_id_b6b11 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_ID_B6B11)

        self._msg_sender.send(msg)

    @property
    def mfr_id(self):
        return self.mfr_id_b0b5 + self.mfr_id_b6b11

    @property
    @functools.lru_cache
    @wait_for("mfr_model_b0b5")
    def mfr_model_b0b5(self):
        """query the mfr_model_b0b5 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_MODEL_B0B5)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_model_b6b11")
    def mfr_model_b6b11(self):
        """query the mfr_model_b6b11 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_MODEL_B6B11)

        self._msg_sender.send(msg)

    @property
    def mfr_model(self):
        return (self.mfr_model_b0b5 + self.mfr_model_b6b11).strip()

    @property
    @functools.lru_cache
    @wait_for("mfr_serial_b0b5")
    def mfr_serial_b0b5(self):
        """query the mfr_serial_b0b5 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_SERIAL_B0B5)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_serial_b6b11")
    def mfr_serial_b6b11(self):
        """query the mfr_serial_b6b11 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_SERIAL_B6B11)

        self._msg_sender.send(msg)

    @property
    def mfr_serial(self):
        return self.mfr_serial_b0b5 + self.mfr_serial_b6b11

    @property
    @functools.lru_cache
    @wait_for("mfr_date_b0b5")
    def mfr_date(self):
        """query the mfr_date_b0b5 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_DATE_B0B5)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_revision_b0b5")
    def mfr_revision(self):
        """query the mfr_revision_b0b5 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_REVISION_B0B5)

        self._msg_sender.send(msg)

    @property
    @functools.lru_cache
    @wait_for("mfr_location_b0b2")
    def mfr_location(self):
        """query the mfr_location_b0b2 of the bic"""
        msg = self._msg_factory(command=pybic.cmd.MFR_LOCATION_B0B2)

        self._msg_sender.send(msg)
